C7Engine/AI/BarbarianAI.cs

```diff
@@ -11,40 +12,89 @@ public class BarbarianAI {
 
 		private ILogger log = Log.ForContext<BarbarianAI>();
 
+		// Threshold at which barbarians take risky actions.
+		// In the future this may vary based on barbarian settings, and perhaps the situation
+		const double COLLEGE_TRY_THRESHOLD = 1.0 / 3;
+
 		public void PlayTurn(Player player, GameData gameData) {
 			if (!player.isBarbarians) {
 				throw new System.Exception("Barbarian AI can only play barbarian players");
 			}
 
-			// Copy unit list into temporary array so we can remove units while iterating.
-			// TODO: We also need to handle units spawned during the loop, e.g. leaders, armies, enslaved units. This is not so much an
-			// issue for the barbs but will be for similar loops elsewhere in the AI logic.
-			foreach (MapUnit unit in player.units.ToArray()) {
-				if (UnitIsFreeToMove(unit)) {
-					while (unit.movementPoints.canMove) {
-						//Move randomly
-						List<Tile> validTiles = unit.unitType.categories.Contains("Sea") ? unit.location.GetCoastNeighbors() : unit.location.GetLandNeighbors();
-						if (validTiles.Count == 0) {
-							//This can happen if a barbarian galley spawns next to a 1-tile lake, moves there, and doesn't have anywhere else to go.
-							log.Warning("WARNING: No valid tiles for barbarian to move to");
-							break;
-						}
-						Tile newLocation = validTiles[GameData.rng.Next(validTiles.Count)];
-						//Because it chooses a semi-cardinal direction at random, not accounting for map, it could get none
-						//if it tries to move e.g. north from the north pole.  Hence, this check.
-						if (newLocation != Tile.NONE) {
-							log.Debug("Moving barbarian at " + unit.location + " to " + newLocation);
-							if (!unit.move(unit.location.directionTo(newLocation))) {
+			BarbarianPlayer barbs = (BarbarianPlayer)player;
+
+			foreach (BarbarianTribe tribe in barbs.getTribes())
+			{
+				// Copy unit list into temporary array so we can remove units while iterating.
+				foreach (MapUnit unit in tribe.GetUnits().ToArray()) {
+					if (UnitIsFreeToMove(unit)) {
+						while (unit.movementPoints.canMove) {
+							//Move randomly
+							List<Tile> validTiles = unit.unitType.categories.Contains("Sea")
+								? unit.location.GetCoastNeighbors()
+								: unit.location.GetLandNeighbors();
+							if (validTiles.Count == 0) {
+								//This can happen if a barbarian galley spawns next to a 1-tile lake, moves there, and doesn't have anywhere else to go.
+								log.Warning("WARNING: No valid tiles for barbarian to move to");
 								break;
 							}
-						} else {
-							//Avoid potential infinite loop.
-							break;
+
+							//Check if there are any undefended units that can be taken!
+							foreach (Tile tile in validTiles) {
+								if (tile.unitsOnTile.Exists(mapUnit => UndefendedUnit(mapUnit))) {
+									bool alive = unit.move(unit.location.directionTo(tile));
+									// TODO: Restructure so we can avoid gotos.
```

Comment:
Right, these are making me nervous. To me this suggests breaking up the nested loops into functions that return a boolean or nullable object and check the return value to decide the next step